end_to_end_tests/common/log_analytics.py

```diff
@@ -78,6 +78,34 @@ def group_by(
         return Table(rows=rows, columns=new_columns)
 
 
+def get_log_analytics_usage_tracking_token_metric_url_by_subscription_id(
+    tenant_id: str,
+    subscription_id: str,
+    resource_group_name: str,
+    app_insights_name: str,
+    start_time: datetime,
+    end_time: datetime,
+):
+    """
+    Build a URL to deep link into the Azure Portal to view APIM token metrics in Log Analytics for the usage tracking policy, split by SubscriptionID.
+    """
+    # ensure start and end time are at the top of the minute so that the Azure Metrics blade can properly render the chart
+    start = start_time.isoformat()[:-12] + "000Z"
+    end = end_time.isoformat()[:-12] + "000Z"
+    time_context = {
+        "absolute": {"startTime": start, "endTime": end},
+        "showUTCTime": False,
+        "grain": 1,
+    }
+    time_context_string = json.dumps(time_context, separators=(",", ":"))
+    encoded_time_context = urllib.parse.quote(time_context_string)
+    # building and encoding the chart_definition JSON object like time_context does not work properly as nested slashes and other special characters are not encoded correctly
+    encoded_chart_definition = f"%7B%22v2charts%22%3A%5B%7B%22metrics%22%3A%5B%7B%22resourceMetadata%22%3A%7B%22id%22%3A%22%2Fsubscriptions%2F{subscription_id}%2FresourceGroups%2F{resource_group_name}%2Fproviders%2FMicrosoft.Insights%2Fcomponents%2F{app_insights_name}%22%7D%2C%22name%22%3A%22customMetrics%2FTotal%20Tokens%22%2C%22aggregationType%22%3A1%2C%22namespace%22%3A%22microsoft.insights%2Fcomponents%2Fkusto%22%2C%22metricVisualization%22%3A%7B%22displayName%22%3A%22Total%20Tokens%22%7D%7D%5D%2C%22title%22%3A%22Sum%20Total%20Tokens%20for%20{app_insights_name}%20by%20Subscription%20ID%22%2C%22titleKind%22%3A1%2C%22visualization%22%3A%7B%22chartType%22%3A2%2C%22legendVisualization%22%3A%7B%22isVisible%22%3Atrue%2C%22position%22%3A2%2C%22hideHoverCard%22%3Afalse%2C%22hideLabelNames%22%3Atrue%7D%2C%22axisVisualization%22%3A%7B%22x%22%3A%7B%22isVisible%22%3Atrue%2C%22axisType%22%3A2%7D%2C%22y%22%3A%7B%22isVisible%22%3Atrue%2C%22axisType%22%3A1%7D%7D%7D%2C%22grouping%22%3A%7B%22dimension%22%3A%22customDimensions%2FSubscription%20ID%22%2C%22sort%22%3A2%2C%22top%22%3A10%7D%7D%5D%7D"
```

Comment:
Can this be converted to a LogAnalytics query that is encoded (#47 moved some prior metrics over)

Reply:
Oops! Those methods are unused now, screwed up the merge thinking they were still used by the usage-tracking e2e test. Great catch! Removed.